### ISSUE TITLE
Parallelize conv_batched

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#145]
This can be done with rayon.

Parallelizing forward is relatively straightforward, as you can just `result.par_iter_mut().zip(...).foreach()` since filters & bias are ref borrowed.

Parallelizing backward is less striaghtforward, but easiest way is to allocate a filters & bias with batch size dimension, and then par_iter_mut() on those two and img_grad. Then at the end of backward up, sum the batched filters & bias into filter_grad & bias_grad.

```rust
pub fn conv2d_par_batched<
    TAPE: 'static + Tape,
    const BATCH_SIZE: usize,
    const IN_CHAN: usize,
    const OUT_CHAN: usize,
    const KERNEL: usize,
    const STRIDE: usize,
    const PADDING: usize,
    const IN_HEIGHT: usize,
    const IN_WIDTH: usize,
>(
    x: Tensor4D<BATCH_SIZE, IN_CHAN, IN_HEIGHT, IN_WIDTH, TAPE>,
    filters: &Tensor4D<OUT_CHAN, IN_CHAN, KERNEL, KERNEL>,
    bias: &Tensor1D<OUT_CHAN>,
) -> Tensor4D<
    BATCH_SIZE,
    OUT_CHAN,
    { (IN_HEIGHT + 2 * PADDING - KERNEL) / STRIDE + 1 },
    { (IN_WIDTH + 2 * PADDING - KERNEL) / STRIDE + 1 },
    TAPE,
> {
    let mut result = Tensor4D::zeros();
    let f_data = filters.data();
    let b_data = bias.data();
    result
        .mut_data()
        .par_iter_mut()
        .zip(x.data().par_iter())
        .for_each(|(r_i, x_i)| {
            conv_forward::<
                IN_CHAN,
                OUT_CHAN,
                KERNEL,
                STRIDE,
                PADDING,
                IN_HEIGHT,
                IN_WIDTH,
                { (IN_HEIGHT + 2 * PADDING - KERNEL) / STRIDE + 1 },
                { (IN_WIDTH + 2 * PADDING - KERNEL) / STRIDE + 1 },
            >(x_i, f_data, b_data, r_i);
        });

    let f = filters.clone();

    let (x, mut tape) = x.split_tape();
    let phantom_filters = filters.phantom();
    let phantom_bias = bias.phantom();
    let phantom_result = result.phantom();
    tape.add_backward_op(move |grads| {
        let (f_grad, b_grad, i_grad, r_grad) =
            grads.muts_and_ref(&phantom_filters, &phantom_bias, &x, &phantom_result);

        let f_data = f.data();
        let mut fs: Box<[[[[[f32; KERNEL]; KERNEL]; IN_CHAN]; OUT_CHAN]; BATCH_SIZE]> =
            Cpu::zeros();
        let mut bs: Box<[[f32; OUT_CHAN]; BATCH_SIZE]> = Cpu::zeros();
        x.data()
            .par_iter()
            .zip(fs.par_iter_mut())
            .zip(bs.par_iter_mut())
            .zip(r_grad.par_iter())
            .zip(i_grad.par_iter_mut())
            .for_each(|item| {
                let ((((x_i, f_grad), b_grad), r_i), i_i) = item;
                conv_backward::<
                    IN_CHAN,
                    OUT_CHAN,
                    KERNEL,
                    STRIDE,
                    PADDING,
                    IN_HEIGHT,
                    IN_WIDTH,
                    { (IN_HEIGHT + 2 * PADDING - KERNEL) / STRIDE + 1 },
                    { (IN_WIDTH + 2 * PADDING - KERNEL) / STRIDE + 1 },
                >(x_i, f_data, r_i, i_i, f_grad, b_grad);
            });

        for i in 0..BATCH_SIZE {
            Cpu::add(f_grad, &fs[i]);
            Cpu::add(b_grad, &bs[i]);
        }
    });
    result.put_tape(tape)
}
```

**Comment by coreylowman** [comment]
Related to #186

**Comment by coreylowman** [comment]
Closing this ticket as its pretty out of date, should be addressed in #186
